Tests refactor

liavkoren opened the pull request:
Hey, Jonas, I still haven't added references to the heroku/AWS credentials, but I figured it's better to (finally) get a PR open to you, in case you had any comments/changes you wanted. This PR adds the artifactCreation logic to the task-succeeds test. I expect this test should be pretty easy to adapt to the task-fails test.

Commit:
rm environmentVariable.js file, creates taskComplete.js and taskFailed.js versions of it.

Commit:
Start of taskComplete test refactor.

Commit:
Gets basic pattern for test working.

Commit:
Full chain of events in taskCompleted test.

Commit:
Cleans up taskComplete.js test

Commit:
Adds artifact creation logic to the integration test.

Commit:
Fixes run-diagnostics

liavkoren commented on diagnostics/docker-worker/taskCompleted.js:
```diff
@@ -0,0 +1,126 @@
+        debug("Getting artifact");
+        if (response.ok) {
+          assert(response.text === testArtifact, "Received artifact does not match expected.");
+          done();
```
I needed to used `done()` here to make sure the callback is hit; not sure if there's a better/more modern way to do this -- I thought initially it might be better to promisify the callback, but a quick googling suggested that might not be a good idea.

jonasfj replied:
There are two approaches here:

``` js
return new Promise(function(accept, reject) {
  request.get(artifactUrl).end(function(err, response){
    if (err) {
      reject(err);
      return;
    }
    accept(response);
  });
}).then(function(response) {
  assert(response.statusCode === 200, "Expected statusCode 200");
  assert(response.text === testArtifact, "Received artifact does not match expected.");
});
```

Or you can use `var request = require('superagent-promise');`.

``` js
return request.get(artifactUrl).end().then(function(response) {
  assert(response.statusCode === 200, "Expected statusCode 200");
  assert(response.text === testArtifact, "Received artifact does not match expected.");
});
```

Note, `'superagent-promise'` is a module that wraps superagent, so that `end()` returns a promise. You still have to list superagent in your `package.json`, because it's relies on the `peerDependency` feature. Hence, you can use `superagent-promise` with whatever version of superagent you like.

jonasfj commented on diagnostics/docker-worker/taskCompleted.js:
```diff
@@ -0,0 +1,126 @@
+        request.put(queueResponse.putUrl)
+        .set('Content-Type', 'application/json')
+        .set("Content-Length", testArtifact.length)
+        .send(testArtifact).end();
```
Hmm... You should probably wait for the request to be done, before you proceed..
If used superagent-promise, then just:

``` js
return request.put(queueResponse.putUrl)
            .set('Content-Type', 'application/json')
            .set("Content-Length", testArtifact.length)
            .send(testArtifact).end();
```

Should do the trick... Otherwise you risk that you try to download the artifact before upload was completed. Naturally that won't work..

jonasfj commented:
Generally, this looks good!

See comments on superagent-promise, it makes using superagent a lot easier... In a promised based flow.

Commit:
Switches taskComplete test from superagent to superagent-promises and removes a potential race condition.

Commit:
Adds an failing task test.

Commit:
Adds a parametrized task success/task fails alternative implementation.

liavkoren commented:
Right! 8c4b7e2 implements your suggestions.  

A few months ago you mentioned wanting a version of this test where the task exists with 1. I've added taskFailed.js to this PR -- it's really the same test as taskCompleted.js, with just the relevant details changed. I've also added an alternative implementation, tasks.js, which parametrizes the test based on sucess/failure. Let me know what your thoughts are about the best implementation approach. Eg, does the failing task case also need to create an artifact?   

You also mentioned want to download and search log files. How do I locate the log file generated by a task run? You included a snippet 

``` javascript

{
   ...
   payload: {
      image: ...
      command: ...
      artifacts: {
         "public/folder/my-artifact-name": {
               type: "file",
               path: "/etc/passwd",
               expires: new Date(/* future */)
         }
      }
   }
}
```

But I haven't been able to figure out from the Docker worker and Queue documentation how you get access to artifacts created within the containers -- would we basically need to run code inside the container which does a PUT to S3?

Anyway, I'm sure this is a lot process all at once. Sorry for the delay followed by the info dump. :ghost: :goberserk: :gift:

jonasfj commented:
We can use the `listArtifacts` api to find the log... or we can just use the name `public/logs/live.log` which is currently hardcoded...
You are right I don't think this is documented... but it's easily evident from the task-inspector:
https://tools.taskcluster.net/task-inspector/#c4Wrjks7QHu8akeNRS8F0Q/0

Regarding your implementation, I think that's a smart way of quickly creating two test cases :)

---

I think we should focus on getting this deployed so we have it running and monitoring health of the system... Then we can always improve further, I suspect just having something is good enough.

jonasfj commented:
Btw, I think I invited you to a team that can edit this repository... Feel free to merge this PR... so we can continue on the road to deploy it :)

Commit:
Removes taskComplet.js and taskFails.js in favour of tasks.js

liavkoren commented:
I just checked the heroku app, looks like this is already deployed. :smile_cat: 

What should I focus on next?
